compiler/rustc_middle/src/mir/query.rs

```diff
@@ -114,13 +114,45 @@ pub struct UnsafetyViolation {
     pub details: UnsafetyViolationDetails,
 }
 
-#[derive(Clone, TyEncodable, TyDecodable, HashStable, Debug)]
+#[derive(Copy, Clone, PartialEq, TyEncodable, TyDecodable, HashStable, Debug)]
+pub enum UnusedUnsafe {
+    /// `unsafe` block contains no unsafe operations
+    // > ``unnecessary `unsafe` block``
+    Unused,
+    // `unsafe` block nested under another (used) `unsafe` block
+    // ``… because it's nested under this `unsafe` block``
+    InUnsafeBlock(hir::HirId),
+    // `unsafe` block nested under `unsafe fn`
+    // ``… because it's nested under this `unsafe fn` ``
+    //
+    // the second HirId here indicates the first usage of the `unsafe` block,
+    // which allows retrival of the LintLevelSource for why that operation would
+    // have been permitted without the block
+    InUnsafeFn(hir::HirId, hir::HirId),
+}
+
+// The `Ord` implementation is relevant, because `SomeDisallowedInUnsafeFn` takes
+// precedence and we want the smallest `HirId`
+#[derive(Copy, Clone, PartialEq, PartialOrd, Eq, Ord, TyEncodable, TyDecodable, HashStable, Debug)]
+pub enum UsedUnsafeBlockData {
+    SomeDisallowedInUnsafeFn,
+    // the HirId here indicates the first usage of the `unsafe` block
+    AllAllowedInUnsafeFn(hir::HirId),
```

Comment:
okay

Reply:
I kind-of derived that by reading this part in the original implementation
```rs
    // The unused unsafe blocks might not be in source order; sort them so that the unused unsafe
    // error messages are properly aligned and the issue-45107 and lint-unused-unsafe tests pass.
    unsafe_unused.sort_by_cached_key(|hir_id| tcx.hir().span(*hir_id));
```

I don’t know how true it is that HirId really corresponds in this way to source order. In any case note that the only usage of this `HirId` here is for providing one instance of where the `unsafe_op_in_unsafe_fn` lint level was defined for one of the unsafe operations in the block. (Typically, it’s going to be the same source for all usages anyways.) There is no strict need for it to be the first one, it’s more about being consistent and not providing more information than necessary. (In particular, changing the `unsafe_op_in_unsafe_fn` lint level for ***any*** of the operations would already make the `unused_unsafe` warning disappear).

Reply:
I mean, just for sanity, it might make sense to somehow determine (i.e. ask somebody) whether or not `HirId` order ***does*** actually correspond (to some degree) to order of appearance in the source code. Or to order of appearance in some other immediate representation, e.g. of the AST, in case that makes a difference.

Reply:
Well, the code above

```
unsafe_unused.sort_by_cached_key(|hir_id| tcx.hir().span(*hir_id));
```

is sorting the _spans_ of the HIR node, not the HirIds themselves, and those do have direct relationship with the appearance order in the source code.

It might be better if in the code below, where the comparison and logic for this invariant is implemented (that is `*entry = cmp::min(*entry, new_usage);`), the spans were compared as well?

Though my understanding of your explanation that it doesn't actually matter that this HirId actually refers to the first `unsafe` use in the source-order, is that correct? I think its okay too and the adjusted comment is clear enough, I guess, what the “first” here really means.

Reply:
>sorting the *spans* of the HIR node, not the HirIds themselves

Oh… I don’t know how I missed that. Perhaps misreading it once, or even misremembering and then never double-checking.

Maybe the more systematic approach would then be to just take the first one appearing in MIR (i.e. just the first one we come across in the MIR traversal)? 🤔 

Indeed it doesn’t really matter, I would appreciate reproducible compilation results (in terms of the set of warnings) though, even in the rare edge cases where this logic actually matters, so some sort of systematic approach is good.

Right, I think I’ll change it to first appearance in MIR then, this is particularly straightforward as it’s just going to be keeping the first operation that’s encountered, making it the cheapest option, too.

Reply:
I’ve changed it now